control-plane-operator/controllers/metrics/metricscontroller.go

```diff
@@ -0,0 +1,208 @@
+package metrics
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/transport"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	promapi "github.com/prometheus/client_golang/api"
+	promv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+)
+
+var (
+	ComponentLabel = "app"
+
+	ComponentMemoryRSS = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "hypershift_component_memory_rss",
+		Help: "Component memory usage in bytes grouped by control plane component.",
+	}, []string{ComponentLabel})
+)
+
+func init() {
+	metrics.Registry.MustRegister(ComponentMemoryRSS)
+}
+
+// MetricsController is responsible for reading metrics from the management cluster
+// about the hosted control plane and transforming and re-exposing them as more
+// targeted metrics suitable for high level SLO analysis (e.g. via CI runs).
+//
+// To do the work, the controller periodically queries the user workload monitoring
+// route and correlates the metrics with pods.
+type MetricsController struct {
+	podCache             cache.Cache
+	namespace            string
+	userWorkloadEndpoint string
+	log                  logr.Logger
+}
+
+// Options configure the metrics controller.
+type Options struct {
+	// Namespace is the namespace of the hosted control plane.
+	Namespace string
+
+	// PodCache should provide access to pods in Namespace.
+	PodCache cache.Cache
```

Comment:
Can we make this a `crclient.Reader` so testing doesn't entail building a fake cache? The cache implements that interface.